numba/interpreter.py

```diff
@@ -242,119 +243,169 @@ def _compute_generator_info(self):
+            combined_liveset |= terminator_liveset
+            # vars that are dead within the block beacuse they are not
+            # propagated to any outgoing blocks
+            intern_set = cur_live_set - combined_liveset
+            intern_dead_map[offset] = intern_set
+            # vars that escape this block
+            escaping_live_set = cur_live_set - intern_set
+            for out_blk, new_live_set in outgoing_live_map.items():
+                # successor should delete the unused escaped vars
+                new_live_set = new_live_set | var_def_map[out_blk]
+                escaping_dead_map[out_blk] |= escaping_live_set - new_live_set
+
+            # if no outgoing blocks
+            if not outgoing_live_map:
+                # insert var used by terminator
+                exit_dead_map[offset] = terminator_liveset
+
+        # Verify that the dead maps cover all live variables
+        all_vars = reduce(operator.or_, live_map.values(), set())
+        inter_dead_vars = reduce(operator.or_, intern_dead_map.values(),
+                                 set())
+        escaping_dead_vars = reduce(operator.or_, escaping_dead_map.values(),
+                                    set())
+        exit_dead_vars = reduce(operator.or_, exit_dead_map.values(), set())
+        dead_vars = (inter_dead_vars | escaping_dead_vars | exit_dead_vars)
+        missing_vars = all_vars - dead_vars
+        if missing_vars:
+            # There are no exit points
+            if not self.cfa.graph.exit_points():
+                # We won't be able to verify this
+                pass
             else:
-                tails.add(b)
+                msg = 'liveness info missing for vars: {0}'.format(missing_vars)
+                raise RuntimeError(msg)
 
-        # Remove temporary entries for infinite loops
-        tails -= infinite_loops
+        return intern_dead_map, escaping_dead_map
 
-        return tails
+    def _patch_var_dels(self, intern_dead_map, escaping_dead_map):
+        """
+        Insert delete in each block
+        """
+        for offset, ir_block in self.blocks.items():
+            # for each intern var, insert delete after the last use
+            intern_set = intern_dead_map[offset]
+            # all non escaping live varibles at the terminator
+            live_set = set(v.name for v in ir_block.terminator.list_vars())
+            live_set &= intern_set
+            delete_pts = []
+            # for each statement in reverse order
+            for stmt in reversed(ir_block.body[:-1]):
+                # intern vars that are used here
+                use_set = set(v.name for v in stmt.list_vars()) & intern_set
+                # used here but not afterwards
+                dead_set = use_set - live_set
+                delete_pts.append((stmt, dead_set))
+                live_set |= use_set
+
+            # insert del after the last uses
+            for stmt, delete_set in delete_pts:
```

Comment:
For the record, this will be O(n²). Perhaps it would be better to rewrite the block's body from scratch.